drivers/docker/driver.go

```diff
@@ -709,6 +745,39 @@ func (d *Driver) containerBinds(task *drivers.TaskConfig, driverConfig *TaskConf
 	return binds, nil
 }
 
+func (d *Driver) recoverPauseContainers() {
+	// On Client restart, we must rebuild the set of pause containers
+	// we are tracking. Basically just scan all containers and pull the ID from
+	// anything that has the Nomad Label and has Name with prefix "/nomad_init_".
+
+	_, dockerClient, err := d.dockerClients()
+	if err != nil {
+		d.logger.Error("failed to recover pause containers", "error", err)
+		return
+	}
+
+	containers, listErr := dockerClient.ListContainers(docker.ListContainersOptions{
+		All: false, // running only
+		Filters: map[string][]string{
+			"label": {dockerLabelAllocID},
+		},
+	})
+	if listErr != nil {
+		d.logger.Error("failed to list pause containers", "error", err)
+		return
+	}
+
+CONTAINER:
+	for _, c := range containers {
+		for _, name := range c.Names {
+			if strings.HasPrefix(name, "/nomad_init_") {
+				d.pauseContainers.add(c.ID)
+				continue CONTAINER
+			}
+		}
+	}
+}
```

Comment:
I think this implementation can lead to us tracking pause containers that have been orphaned by their task. For example:
- the client stops
- the task fails, but the pause container is left running
- the client restarts
- the task cannot be recovered
- the pause container will be picked up here and tracked

We currently have the `RecoverTask` method in this file. Could we try to find the pause container for a task that's been successfully recovered there, so that we only track the pause containers for live tasks?

Reply:
I'm pretty sure this implementation is correct - a pause container is associated with an allocation, not a task. If a task cannot be recovered the alloc is replaced, and the destruction of the alloc will call [DestroyNetwork](https://github.com/hashicorp/nomad/blob/main/client/allocrunner/network_hook.go#L185) the implementation of which now purges the pause container associated with the alloc from the tracking store.

Testing described steps, 

<details><summary>job file</summary>

```hcl
job "bug" {

  group "cache" {
    network {
      mode = "bridge"
      port "db" {
        to = 6379
      }
    }
    
    task "sleep" {
      driver = "docker"
      
      restart {
        attempts = 0
        mode = "fail"
      }
      
      config {
        image = "bash:5"
        command = "sleep"
        args = ["infinity"]
      }
      
      resources {
        cpu = 100
        memory = 32
      }
    }

    task "redis" {
      driver = "docker"

      config {
        image          = "redis:7"
        ports          = ["db"]
        auth_soft_fail = true
      }

      resources {
        cpu    = 100
        memory = 128
      }
    }
  }
}
```
</details>

```
Work/tickets/docker-infra 
➜ docker ps -a
CONTAINER ID   IMAGE                                      COMMAND                  CREATED          STATUS          PORTS     NAMES
5a7c52cb1854   redis:7                                    "docker-entrypoint.s…"   30 seconds ago   Up 29 seconds             redis-9b58cff0-d8ad-b1b0-186f-adcc2b53ed17
c4af36a0a5fa   bash:5                                     "docker-entrypoint.s…"   32 seconds ago   Up 31 seconds             sleep-9b58cff0-d8ad-b1b0-186f-adcc2b53ed17
86b102d2d006   gcr.io/google_containers/pause-amd64:3.1   "/pause"                 33 seconds ago   Up 32 seconds             nomad_init_9b58cff0-d8ad-b1b0-186f-adcc2b53ed17

Work/tickets/docker-infra 
➜ sudo pkill -9 nomad

Work/tickets/docker-infra 
➜ docker stop c4
c4

Work/tickets/docker-infra 10s
➜ docker ps -a
CONTAINER ID   IMAGE                                      COMMAND                  CREATED              STATUS                        PORTS     NAMES
5a7c52cb1854   redis:7                                    "docker-entrypoint.s…"   About a minute ago   Up About a minute                       redis-9b58cff0-d8ad-b1b0-186f-adcc2b53ed17
c4af36a0a5fa   bash:5                                     "docker-entrypoint.s…"   About a minute ago   Exited (137) 10 seconds ago             sleep-9b58cff0-d8ad-b1b0-186f-adcc2b53ed17
86b102d2d006   gcr.io/google_containers/pause-amd64:3.1   "/pause"                 About a minute ago   Up About a minute                       nomad_init_9b58cff0-d8ad-b1b0-186f-adcc2b53ed17

Work/tickets/docker-infra 
➜ # start nomad


Work/tickets/docker-infra 
➜ docker ps -a
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
```

Reply:
Ah, I forgot that we call `DestroyNetwork` if we failed to recover. That works :+1: